mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp

```diff
@@ -908,8 +908,9 @@ void CallOp::build(OpBuilder &builder, OperationState &state, TypeRange results,
 
 void CallOp::build(OpBuilder &builder, OperationState &state, TypeRange results,
                    FlatSymbolRefAttr callee, ValueRange args) {
+  auto fargs = callee ? args : args.drop_front();
```

Comment:
I wonder if it may be better to have an assert that the `callee` is present here, and instead provide a separate builder, which does not take a `callee` argument at all and have that construct an indirect call.

This should be less confusing as the above builders that take a `callee` arg _always_ result in a direct call.

Reply:
But if there is a lot of code that depends on this behaviour then I guess this works as well.

Reply:
I very much agree with Ivan that there should be an assert that checks that `callee` is non-null and possibly an alternative builder signature that does not take a `callee`. Alternatively, the function type could also be computed outside of the builder. However, I think it is fine to add a new builder signature that does this for us.

Reply:
well, let's have one more round :) And I'll do the fix!

well, I have some doubts about the idea to pass a callee as a value ( or I just misunderstood something :) ) The reason is that everywhere the behavior is assumed as it stated in the docs:
with the first operand being a callee in the operand list in the case of indirect call. So passing a callee separately as a value may be another source of confusion - if the operands must contain the callee in this case or not. Let's just add a builder that doesn't have a callee as an argument - and it will be a true indirect call!

And one more thought about: the fact we don't want to check the attribute here means that it need to be checked in the user code - we just move the problem from this place to another. You can consider the ClangIR [code](https://github.com/llvm/clangir/blob/main/clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp#L880l) that lowers CIR to MLIR. So choosing the proper builder would cause adding extra checks in there. 

But of course it's up to you to decide. I would say that the crox of the problem is that we probably need two different operations for direct and indirect calls - in this case no confusion would happen. 

So what's the plan?) What do I need to do? 

@ivanradanov  @gysit

Reply:
The docs refer to the operands of the MLIR op, and not the builder. The builders that take in a `callee` that specify a function using a symbol should result in a direct call IMO.

> Let's just add a builder that doesn't have a callee as an argument - and it will be a true indirect call!

We already have builders do not take callee as an argument, but one needs to provide the callee type instead. This is because there is no way to infer whether the callee is variadic or not in the case of an indirect call.

In fact, I think that CIR code is wrong when the indirect callee is variadic (unless CIR has a different CallOp for variadic functions?)

If you want to retain the current functionality (which is wrong if you have indirect calls to variadic), I believe  you can use the default builder (https://github.com/llvm/llvm-project/blob/d8ddcae547e782a4765783c62747381624f076d4/mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td#L625) and pass in nullptr as the callee and callee_type.

Reply:
So I think the change that is needed upstream is to just add an assert that callee is not nullptr. This will make it explicit that that builder will create a direct call.

Reply:
> This is because there is no way to infer whether the callee is variadic or not in the case of an indirect call.

Good point... didn't consider that the callee type is actually only relevant for variadics. Then using the builder that takes the callee type makes most sense.

> So I think the change that is needed upstream is to just add an assert that callee is not nullptr. This will make it explicit that that builder will create a direct call.

Agreed. Is there actually a reason to set the callee type in the builder that the revision modifies. After all, the builder does not work for variadic functions. So just setting the calle_type to nullptr may also be an option?

Reply:
So, I added an assert in the builder.   And I think we can use [this](https://github.com/llvm/llvm-project/blob/d8ddcae547e782a4765783c62747381624f076d4/mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td#L637l) builder to construct indirect calls

>  Is there actually a reason to set the callee type in the builder that the revision modifies. After all, the builder does not work for variadic functions. So just setting the calle_type to nullptr may also be an option?

Well, in this case we'll break the `getCalleeFunctionType` function, as far as I see

@ivanradanov @gysit

Reply:
`getCalleeFunctionType` seems to be able to work with a `callee_type` set to null (there is an if else specifically for this as far as I can see). It essentially takes the function arguments and results to build the correct type for a non-variadic function. However, I think the question is not relevant for this revision and it does not hurt to set the `callee_type` either.

Reply:
well, the next problem will arise here: once the `callee_type` is not set,  the function type will be inferred from  the operands, and again, with a pointer to the function itself at the first place.

Reply:
```
Operation::operand_range CallOp::getArgOperands() {
  return getOperands().drop_front(getCallee().has_value() ? 0 : 1);
}
```

We drop the first operand when it is not set so it should be inferred correctly.

Reply:
Let's land this then? @gitoleg do you have commit rights? Otherwise, one of us will land.

Reply:
@gysit  
no rights :) 

> We drop the first operand when callee_type is not set so it should be inferred correctly.
@ivanradanov  yes, you're right!

Reply:
ok, I am traveling today. Will land once I am back (@ivanradanov feel free to go ahead if you are around). The title and description of the PR is somewhat outdated but I think I can adapt it when landing.